scripts/check_compliance.py

```diff
@@ -91,6 +91,44 @@ def run(self):
         """
         pass
 
+    def error(self, msg):
+        """
+        Signals a problem with running the test, with message 'msg'.
+
+        Overrides fail() and skip() if those have been called earlier.
+
+        Note: Does not raise an exception.
+        """
+        self.case.result = Error(self._name + " error", "error")
+        self.case.result._elem.text = msg
+
+    def skip(self, msg):
+        """
+        Signals that the test should be skipped, with message 'msg'.
+
+        Overrides fail() and error() if those have been called earlier.
+
+        Note: Does not raise an exception.
+        """
+        self.case.result = Skipped(self._name + " skipped", "skipped")
+        self.case.result._elem.text = msg
+
+    def fail(self, msg):
```

Comment:
How about something like this:

```python

def addFailure(...): # or any other different looking name to highlight the difference with single-use .skip() and error()

if self.case.results and not isinstance(Failure):
   raise "something terrible happened"

if not self.case.results:
     self.case.results = Failure(...)

self.case.result._elem.text += msg + "\n\n"
```
"early return" coding style https://lwn.net/Articles/723489/

Reply:
The ability to override a failure with a later skipped, etc., is deliberate (and documented), for flexibility. Dunno if it's that useful in practice though.

Had this at first (`text` starts out as `None`), but those redundant `"\n\n"` for a single failure bothered me. :)

```python
if not isinstance(self.case.result, Failure):
    # If the result is not a Failure (usually because the result hasn't
    # been set yet), then make it one
    self.case.result = Failure(self._name + " issues", "failure")
    self.case.result._elem.text = ""

# If there are multiple Failures, concatenate their messages
self.case.result._elem.text += "\n\n" + msg
```

Reply:
> The ability to override a failure with a later skipped, etc., is deliberate (and documented), 

Wait... so now reviewers are supposed to look at not just code but also pydoc carefully added by authors? :-/

> for flexibility. Dunno if it's that useful in practice though.

Considering it silently drops information I looks pretty harmful to me... Have you spotted any actual usage?

How about logging.warning() then? It would not change the current behavior and just flag (very..) suspicious use cases.

> though it only shows up in compliance.xml

Errr.. then who cares?

Reply:
> Wait... so now reviewers are supposed to look at not just code but also pydoc carefully added by authors? :-/

Should always document functions IMO (and read docs). Often difficult to know what's intentional and what isn't otherwise.

> Considering it silently drops information I looks pretty harmful to me... Have you spotted any actual usage?

Nope.

I wonder if it's that bad in practice though. Feels like the worst thing that could happen would be that someone would have to run CI twice. Hard to even come up with cases where that'd happen, and it'd probably be obvious.

If someone wants to say "oh wait, skip this test" after adding some failures, then let them, imo. Keeps the code short and simple.

> 
> How about logging.warning() then? It would not change the current behavior and just flag (very..) suspicious use cases.

Don't think it'll ever be a problem in practice. Better to keep it short.

Reply:
> > though it only shows up in compliance.xml
> 
> Errr.. then who cares?

My OCD. Changed.

Reply:
> If someone wants to say "oh wait, skip this test" after adding some failures, then let them, imo

I can't think of why you'd want to skip() *after* adding failures or errors - or any other similar use case.

+1 except for this.

> Feels like the worst thing that could happen would be that someone would have to run CI twice. 

Sorry I don't get this.